File: 0415/SoohyunKim/1135_뉴스전하기.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const N = +input[0];
const list = input[1].split(' ').map(Number);

const company = Array.from({ length: N }, () => []);
// i번 직원의 직속 부하 목록
for (let i = 1; i < N; i++) {
  const boss = list[i];
  company[boss].push(i); 
}

function dfs(manager) {
  const callTimes = [];

  for (const m of company[manager]) {
    callTimes.push(dfs(m));
  }

  // 전화 시간이 긴 부하부터 처리
  callTimes.sort((a, b) => b - a);
  let maxTime = 0;
  for (let i = 0; i < callTimes.length; i++) {
    // 부하 전달 시간 + 대기시간 i + 전화 시간 1 
    maxTime = Math.max(maxTime, callTimes[i] + i + 1);
  }

  return maxTime;
}

console.log(dfs(0));

File: 0415/SeoyeonHong/1135 뉴스 전하기.py
N = int(input())

num = list(map(int, input().split()))
graph = [[] for _ in range(N)]
for i in range(1, N):
    graph[num[i]].append(i)

def dfs(n):
    if not graph[n]:
        return 0
    
    times = []
    for child in graph[n]:
        times.append(dfs(child))
    
    times.sort(reverse=True) # 오름차순 정렬

    max_time = 0
    for i, t in enumerate(times):
        max_time = max(max_time, t + i + 1)
    
    return max_time

print(dfs(0))